lazy_dataset/core.py

```diff
@@ -2720,27 +2717,34 @@ def _get_memory_size(keep_mem_free):
             import humanfriendly
             return humanfriendly.parse_size(keep_mem_free, binary=True)
 
+    def check(self):
+        if self._keep_mem_free is None:
+            return True
+
+        import psutil
+        if psutil.virtual_memory().available <= self._keep_mem_free:
+            # Return without writing to cache if there is not enough
+            # free memory
+            import warnings
+            warnings.warn(
+                'Max capacity of the in-memory cache is reached. '
+                'The remaining data will not be cached.'
+            )
+            return False
+        return True
+
     def __getitem__(self, item):
         if isinstance(item, str):
             item = self.keys().index(item)
 
         if isinstance(item, numbers.Integral):
-            if item not in self.cache:
-                # Check if we have enough free memory
-                if self._keep_mem_free is not None:
-                    import psutil
-                    if psutil.virtual_memory().available <= self._keep_mem_free:
-                        # Return without writing to cache if there is not enough
-                        # free memory
-                        import warnings
-                        warnings.warn(
-                            'Max capacity of the in-memory cache is reached. '
-                            'The remaining data will not be cached.'
-                        )
-                        return self.input_dataset[item]
-
-                self.cache[item] = self._serialize(self.input_dataset[item])
-            return self._deserialize(self.cache[item])
+            if item in self._cache:
```

Comment:
I think the default case will be later that the key is in the cache. Is EAFP code style then better (less overhead)?

> EAFP
> Easier to ask for forgiveness than permission. This common Python coding style assumes the existence of valid keys or attributes and catches exceptions if the assumption proves false. This clean and fast style is characterized by the presence of many try and except statements. The technique contrasts with the LBYL style common to many other languages such as C.
> https://docs.python.org/3.4/glossary.html#:~:text=Easier%20to%20ask%20for%20forgiveness,many%20try%20and%20except%20statements.

Reply:
Done, switched to EAFP. This (very) slightly speeds up things if the key is in the cache.